# ESTADÍSTICA: SOLUCIONES A EJERCICIOS

Estos ejercicios te permitirán repasar y consolidar lo que has aprendido en el módulo de estadística.

No hagas los ejercicios hasta que hayas estudiado el contenido de su notebook.

IMPORTANTE: el módulo de estadística pretende enseñarte los conceptos de estadística necesarios para data science. El foco no está en usar uno u otro paquete. Por tanto te recomiendo que para resolver los ejercicios no te compliques y uses exactamente los mismos paquetes y funciones que usamos en el módulo (copiar-pegar si quieres) como te digo lo importante es consolidar los conceptos que tendremos que usar en el resto del curso.

Sobre gráficos no haremos ningún ejercicio ya que tenemos un módulo completo dedicado a visualización.

Instrucciones:

* Crea una copia de este notebook antes de escribir nada. De esa forma respetarás el original para volver a hacerlo
* Cada ejercicio tiene unas instrucciones de lo que debes hacer en él
* Bajo las instrucciones estará una celda vacía, aquí es donde tú debes escribir la solución
* Y bajo esta última celda verás que hay otras dos celdas. La primera está vacía y no la debes tocar. Es para que tengas siempre a mano la solución al ejercicio, ya que esta no va a cambiar

In [1]:
#Te dejo creada la importación de paquetes
#Simplemente ejecuta esa celda para que se cargen
import pandas as pd
import numpy as np
import statistics
import scipy as sp
import seaborn as sns
import random
import math
from statsmodels.stats.proportion import proportions_ztest

Vamos a trabajar con un dataset de ventas de productos para hacer los ejerecicios.

Te lo dejo ya cargado como df.

De momento por sencillez le quitaremos todos los nulos, aprendemos mucho más sobre como gestionarlos mejor más adelante en el programa.

In [2]:
df = pd.read_csv('../../00_Datasets/HistoricoVentas.csv', sep = ',', index_col = 0)
df.dropna(inplace = True)
df

Product_Code Warehouse Product_Category        Date  Order_Demand
1      Product_1517    Whse_J     Category_019  2014/11/12         200.0
2      Product_1987    Whse_J     Category_005   2014/2/28         700.0
3      Product_0187    Whse_J     Category_007    2016/1/8           3.0
4      Product_1583    Whse_J     Category_005   2014/3/20         100.0
5      Product_0110    Whse_J     Category_032   2013/7/19         360.0
...             ...       ...              ...         ...           ...
49996  Product_1458    Whse_J     Category_019    2014/6/2         200.0
49997  Product_2082    Whse_A     Category_009  2015/10/21         590.0
49998  Product_0038    Whse_C     Category_005   2015/7/10        5000.0
49999  Product_1393    Whse_J     Category_019   2012/1/17        9000.0
50000  Product_0023    Whse_J     Category_005   2016/6/16         300.0

[49168 rows x 5 columns]

## Ejercicio:

Haz un conteo de frecuencias de cuantas ventas hay de cada categoría de producto.

In [3]:
#No tocar, es para que veas la solución

Product_Category
Category_019    22418
Category_005     4754
Category_001     4634
Category_007     3914
Category_021     2414
Category_006     1686
Category_028     1506
Category_011     1164
Category_015     1045
Category_024      974
Category_009      947
Category_026      708
Category_030      610
Category_032      456
Category_022      403
Category_023      367
Category_018      234
Category_003      185
Category_013      177
Category_020      156
Category_031       85
Category_008       81
Category_033       76
Category_012       47
Category_010       42
Category_029       37
Category_017       23
Category_004       14
Category_016        5
Category_025        3
Category_002        2
Category_027        1
Name: count, dtype: int64

## Ejercicio:

¿Cual es la categoría de producto más vendida? (saca la moda).

In [4]:
#No tocar, es para que veas la solución

'Category_019'

## Ejercicio:

¿Cuantas ventas ha habido de cada categoría de producto en cada almacén (Warehouse)? (haz una tabla cruzada)

Consejo: como hay muchas más categorías que almacenes pon las categorías como filas. 

In [5]:
#No tocar, es para que veas la solución

Warehouse         Whse_A  Whse_C  Whse_J  Whse_S
Product_Category                                
Category_001           8     109    4399     118
Category_002           0       0       0       2
Category_003          96       0      62      27
Category_004           0       0       0      14
Category_005          74     169    4286     225
Category_006         190      41    1303     152
Category_007         706     271    2644     293
Category_008           4       1      76       0
Category_009         359      38     239     311
Category_010          31       0      10       1
Category_011          43      22    1056      43
Category_012           9       5      17      16
Category_013          51       0     109      17
Category_015         238     130     476     201
Category_016           3       0       0       2
Category_017           8       0      15       0
Category_018         117       0      89      28
Category_019        2954     829   17575    1060
Category_020           0       5       0     151
Category_021         634      77    1464     239
Category_022          39       0      79     285
Category_023          20     101     159      87
Category_024         335      97     360     182
Category_025           0       0       0       3
Category_026         171       0     513      24
Category_027           1       0       0       0
Category_028         542       0     887      77
Category_029          37       0       0       0
Category_030           0      96     100     414
Category_031          43       0      42       0
Category_032           0      10     295     151
Category_033           0       0      76       0

## Ejercicio:

Ahora haz lo mismo pero que salga en porcentaje de las categorías de producto(que las filas sumen 100%)

In [6]:
#No tocar, es para que veas la solución

Warehouse           Whse_A    Whse_C    Whse_J    Whse_S
Product_Category                                        
Category_001      0.001726  0.023522  0.949288  0.025464
Category_002      0.000000  0.000000  0.000000  1.000000
Category_003      0.518919  0.000000  0.335135  0.145946
Category_004      0.000000  0.000000  0.000000  1.000000
Category_005      0.015566  0.035549  0.901557  0.047329
Category_006      0.112693  0.024318  0.772835  0.090154
Category_007      0.180378  0.069239  0.675524  0.074859
Category_008      0.049383  0.012346  0.938272  0.000000
Category_009      0.379092  0.040127  0.252376  0.328405
Category_010      0.738095  0.000000  0.238095  0.023810
Category_011      0.036942  0.018900  0.907216  0.036942
Category_012      0.191489  0.106383  0.361702  0.340426
Category_013      0.288136  0.000000  0.615819  0.096045
Category_015      0.227751  0.124402  0.455502  0.192344
Category_016      0.600000  0.000000  0.000000  0.400000
Category_017      0.347826  0.000000  0.652174  0.000000
Category_018      0.500000  0.000000  0.380342  0.119658
Category_019      0.131769  0.036979  0.783968  0.047283
Category_020      0.000000  0.032051  0.000000  0.967949
Category_021      0.262635  0.031897  0.606462  0.099006
Category_022      0.096774  0.000000  0.196030  0.707196
Category_023      0.054496  0.275204  0.433243  0.237057
Category_024      0.343943  0.099589  0.369610  0.186858
Category_025      0.000000  0.000000  0.000000  1.000000
Category_026      0.241525  0.000000  0.724576  0.033898
Category_027      1.000000  0.000000  0.000000  0.000000
Category_028      0.359894  0.000000  0.588977  0.051129
Category_029      1.000000  0.000000  0.000000  0.000000
Category_030      0.000000  0.157377  0.163934  0.678689
Category_031      0.505882  0.000000  0.494118  0.000000
Category_032      0.000000  0.021930  0.646930  0.331140
Category_033      0.000000  0.000000  1.000000  0.000000

## Ejercicio:

¿Está relacionada la venta de ciertas categorías de productos con ciertos almacenes? ¿o por el contrario se venden igual en todos?

Empieza creando la tabla de contingencia.

In [8]:
#No tocar, es para que veas la solución

Warehouse         Whse_A  Whse_C  Whse_J  Whse_S
Product_Category                                
Category_001           8     109    4399     118
Category_002           0       0       0       2
Category_003          96       0      62      27
Category_004           0       0       0      14
Category_005          74     169    4286     225
Category_006         190      41    1303     152
Category_007         706     271    2644     293
Category_008           4       1      76       0
Category_009         359      38     239     311
Category_010          31       0      10       1
Category_011          43      22    1056      43
Category_012           9       5      17      16
Category_013          51       0     109      17
Category_015         238     130     476     201
Category_016           3       0       0       2
Category_017           8       0      15       0
Category_018         117       0      89      28
Category_019        2954     829   17575    1060
Category_020           0       5       0     151
Category_021         634      77    1464     239
Category_022          39       0      79     285
Category_023          20     101     159      87
Category_024         335      97     360     182
Category_025           0       0       0       3
Category_026         171       0     513      24
Category_027           1       0       0       0
Category_028         542       0     887      77
Category_029          37       0       0       0
Category_030           0      96     100     414
Category_031          43       0      42       0
Category_032           0      10     295     151
Category_033           0       0      76       0

## Ejercicio:
Ahora calcula el pvalor del chicuadrado para ver si existe relación o no al 95 de NC. 
(si el pvalor es menor que 0.05 es que sí hay relación entre las dos variables).

In [9]:
#No tocar, es para que veas la solución

np.float64(0.0)

## Ejercicio:

Usando la variable ventas (Order_Demand) calcula la mediana de las ventas

In [11]:
#No tocar, es para que veas la solución

mediana: 300.00


## Ejercicio:

Ahora calcula la desviación típica y la varianza.

In [12]:
#No tocar, es para que veas la solución

desviación típica: 32574.04
varianza: 1061068179.68


## Ejercicio:

Voy a crear para ti un nuevo dataset, llamado dfr (de df recortado) con una muestra más manejable que incluya solo las ventas de la categoría de producto 019.

(más adelante aprenderás cómo hacer todo esto, de momento te lo dejo hecho)

In [13]:
dfr = df[df.Product_Category == 'Category_019']
dfr

Product_Code Warehouse Product_Category        Date  Order_Demand
1      Product_1517    Whse_J     Category_019  2014/11/12         200.0
6      Product_1463    Whse_J     Category_019   2013/12/5        5000.0
8      Product_1472    Whse_J     Category_019   2015/8/24         300.0
11     Product_1458    Whse_J     Category_019    2015/9/2        4000.0
12     Product_1365    Whse_J     Category_019    2014/4/1         300.0
...             ...       ...              ...         ...           ...
49991  Product_1381    Whse_J     Category_019   2014/9/24       10000.0
49992  Product_1513    Whse_A     Category_019    2013/6/4        1000.0
49994  Product_1381    Whse_J     Category_019   2014/6/24        1000.0
49996  Product_1458    Whse_J     Category_019    2014/6/2         200.0
49999  Product_1393    Whse_J     Category_019   2012/1/17        9000.0

[22418 rows x 5 columns]

## Ejercicio:

Calcula el error típico de la variable Order_Demand a patir de dfr.

In [14]:
#No tocar, es para que veas la solución

np.float64(311.1921924007865)

## Ejercicio:

Calcula el margen de error a un 95,5% de confianza.

In [15]:
#No tocar, es para que veas la solución

np.float64(622.384384801573)

## Ejercicio:

Calcula el límite inferior y superior del intervalo de confianza a un 95,5% de confianza. (recuerda que deberás calcular también la media)

In [16]:
#No tocar, es para que veas la solución

media: 9137.23
limite inferior: 8514.84
limite superior: 9759.61


## Ejercicio:

Calcula qué porcentaje de los registros recogidos en dfr corresponden al producto 1359.

In [17]:
#No tocar, es para que veas la solución

Product_Code
Product_1359    0.034570
Product_1295    0.022036
Product_1378    0.021902
Product_1286    0.019582
Product_1382    0.018780
                  ...   
Product_1625    0.000045
Product_0836    0.000045
Product_2151    0.000045
Product_1907    0.000045
Product_1334    0.000045
Name: proportion, Length: 521, dtype: float64

## Ejercicio:

Calcula el intervalo de confianza de esa proporción a un 99,7% de confianza.

(Consulta los apuntes para refrescar la fórmula del error típico en proporciones)

In [18]:
#No tocar, es para que veas la solución

proporción: 0.035
limite inferior: 0.031
limite superior: 0.038


## Ejercicio:

Recordemos que la media de ventas en dfr es de 9137$. Pero eso es la muestra que hemos podido recoger.

Sin embargo el director comercial nos dice que realmente la media está en 9500$.

¿Nos podemos creer lo que dice con un nivel de seguridad del 95%?

In [19]:
#No tocar, es para que veas la solución
#Paso 1: establecemos la hipótesis nula


#Paso 2: definimos un alpha 


#Paso 3:  calculamos el pvalor según el estadístico de contraste elegido (t sobre una muestra)


np.float64(0.24372947061475375)

## Ejercicio:

Saca de nuevo la tabla cruzada de la categoría de productos cruzada con los almacenes (Warehouse), pero en este caso saca los porcentajes verticales (que las columnas sumen el 100%)

In [20]:
#No tocar, es para que veas la solución

Warehouse           Whse_A    Whse_C    Whse_J    Whse_S
Product_Category                                        
Category_001      0.001192  0.054473  0.121081  0.028620
Category_002      0.000000  0.000000  0.000000  0.000485
Category_003      0.014301  0.000000  0.001707  0.006549
Category_004      0.000000  0.000000  0.000000  0.003396
Category_005      0.011023  0.084458  0.117971  0.054572
Category_006      0.028303  0.020490  0.035865  0.036866
Category_007      0.105169  0.135432  0.072775  0.071065
Category_008      0.000596  0.000500  0.002092  0.000000
Category_009      0.053478  0.018991  0.006578  0.075431
Category_010      0.004618  0.000000  0.000275  0.000243
Category_011      0.006405  0.010995  0.029066  0.010429
Category_012      0.001341  0.002499  0.000468  0.003881
Category_013      0.007597  0.000000  0.003000  0.004123
Category_015      0.035454  0.064968  0.013102  0.048751
Category_016      0.000447  0.000000  0.000000  0.000485
Category_017      0.001192  0.000000  0.000413  0.000000
Category_018      0.017429  0.000000  0.002450  0.006791
Category_019      0.440042  0.414293  0.483747  0.257094
Category_020      0.000000  0.002499  0.000000  0.036624
Category_021      0.094444  0.038481  0.040296  0.057967
Category_022      0.005810  0.000000  0.002174  0.069124
Category_023      0.002979  0.050475  0.004376  0.021101
Category_024      0.049903  0.048476  0.009909  0.044143
Category_025      0.000000  0.000000  0.000000  0.000728
Category_026      0.025473  0.000000  0.014120  0.005821
Category_027      0.000149  0.000000  0.000000  0.000000
Category_028      0.080739  0.000000  0.024414  0.018676
Category_029      0.005512  0.000000  0.000000  0.000000
Category_030      0.000000  0.047976  0.002752  0.100412
Category_031      0.006405  0.000000  0.001156  0.000000
Category_032      0.000000  0.004998  0.008120  0.036624
Category_033      0.000000  0.000000  0.002092  0.000000

Fíjate en la penetración de ventas de la categoría 006 en los almacenes A y S.

Que es 0.028 y 0.036 respectivamente.

¿Podríamos pensar que esa diferencia es real y esa categoría se vende más en el S? ¿O es simplemente fruto de nuestros datos y no representa una diferencia real?

Queremos estar muy seguros, así que contrástalo al 99% de confianza.

In [21]:
#No tocar, es para que veas la solución
#Primero sacaremos la tabla cruzada en absoluto y con los marginales
#Ya que lo necesitaremos para lo que le tenemos que pasar a la función


Warehouse         Whse_A  Whse_C  Whse_J  Whse_S    All
Product_Category                                       
Category_001           8     109    4399     118   4634
Category_002           0       0       0       2      2
Category_003          96       0      62      27    185
Category_004           0       0       0      14     14
Category_005          74     169    4286     225   4754
Category_006         190      41    1303     152   1686
Category_007         706     271    2644     293   3914
Category_008           4       1      76       0     81
Category_009         359      38     239     311    947
Category_010          31       0      10       1     42
Category_011          43      22    1056      43   1164
Category_012           9       5      17      16     47
Category_013          51       0     109      17    177
Category_015         238     130     476     201   1045
Category_016           3       0       0       2      5
Category_017           8       0      15       0     23
Category_018         117       0      89      28    234
Category_019        2954     829   17575    1060  22418
Category_020           0       5       0     151    156
Category_021         634      77    1464     239   2414
Category_022          39       0      79     285    403
Category_023          20     101     159      87    367
Category_024         335      97     360     182    974
Category_025           0       0       0       3      3
Category_026         171       0     513      24    708
Category_027           1       0       0       0      1
Category_028         542       0     887      77   1506
Category_029          37       0       0       0     37
Category_030           0      96     100     414    610
Category_031          43       0      42       0     85
Category_032           0      10     295     151    456
Category_033           0       0      76       0     76
All                 6713    2001   36331    4123  49168

In [22]:
#No tocar, es para que veas la solución
#En este caso es un contrate de proporciones sobre 2 muestras diferentes
#Una la del almaceén A y otra la del S

#Paso 1: establecemos la hipótesis nula
#h0: la propoción de ventas de la categoría 006 es realmente la misma en los almacenes A y S

#Paso 2: definimos un alpha 


#Paso 3:  calculamos el pvalor según el estadístico de contraste elegido (z sobre muestras independientes)


#lo pasamos a array porque así lo pide la función


0.013
